File: src/main/java/com/test/util/MessageDecorator.java
package com.test.util;

import com.test.domain.Message;

/**
 * Created by alvarovilaplanagarcia on 21/04/15.
 */
public class MessageDecorator {
    final TimeProvider timeProvider;

    public MessageDecorator(final TimeProvider timeProvider) {
        this.timeProvider = timeProvider;
    }

    public String decorate(Message message) {
        long seconds = (timeProvider.currentTime().getMillis() - message.getCreated().getMillis()) / 1000;
        if (seconds < 60) {
            return message.getMessage() + " (" + seconds + " seconds ago)";
        } else {
            final long minutes = seconds / 60;
            if (minutes < 60) {
                return message.getMessage() + " (" + minutes + " minutes ago)";
            } else {
                final long hours = minutes / 60;
                if (hours < 24) {
                    return message.getMessage() + " (" + hours + " hours ago)";
                }
            }
        }
        return message.getMessage();
    }
}

File: scala/src/main/scala/com/test/util/MessageDecorator.scala
package com.test.util

import com.test.domain.Message
import org.joda.time.DateTime

trait TimeProvider {
  def currentTime: DateTime
}

object TimeProvider extends TimeProvider {
  override def currentTime: DateTime = DateTime.now()
}

class MessageDecorator {
  def decorateWithUser(message: Message, currentTime: DateTime): String = s"${message.user} - ${decorate(message, currentTime)}"


  def decorate(m: Message, currentTime: DateTime): String = {
    val seconds = (currentTime.getMillis() - m.created.getMillis()) / 1000
    val minutes = seconds / 60
    val hours = minutes / 60
    (seconds, minutes, hours) match {
      case (s, _, _) if s < 60 => s"${m.message} ($seconds seconds ago)"
      case (_, min, _) if min < 60 => s"${m.message} ($minutes minutes ago)"
      case (_, _, h) if h < 24 => s"${m.message} ($hours hours ago)"
      case (_, _, _) => m.message

    }
  }

}
